### PR TITLE
New Data Source: aws_lambda_function

### PR DESCRIPTION
**Pull request by loivis**
To get information of lambda function.

Attributes are returned from read function of `aws_lambda_function` resource. Most of the attributes for resource are available for data source. `tags` is one exception though because it doesn't exist if `qualifier` is specified.

### CODE REVIEW

**Comment by loivis** [comment]
Not sure why acceptance test for resource lambda function fails sometime and retry will just pass.
```
⎇  make testacc TEST=./aws TESTARGS='-run=TestAccAWSLambdaFunction'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSLambdaFunction -timeout 120m
=== RUN   TestAccAWSLambdaFunction_importLocalFile
--- PASS: TestAccAWSLambdaFunction_importLocalFile (40.69s)
=== RUN   TestAccAWSLambdaFunction_importLocalFile_VPC
--- PASS: TestAccAWSLambdaFunction_importLocalFile_VPC (37.28s)
=== RUN   TestAccAWSLambdaFunction_importS3
--- PASS: TestAccAWSLambdaFunction_importS3 (38.14s)
=== RUN   TestAccAWSLambdaFunction_basic
--- PASS: TestAccAWSLambdaFunction_basic (38.03s)
=== RUN   TestAccAWSLambdaFunction_concurrency
--- PASS: TestAccAWSLambdaFunction_concurrency (53.49s)
=== RUN   TestAccAWSLambdaFunction_concurrencyCycle
--- PASS: TestAccAWSLambdaFunction_concurrencyCycle (65.14s)
=== RUN   TestAccAWSLambdaFunction_updateRuntime
--- PASS: TestAccAWSLambdaFunction_updateRuntime (55.47s)
=== RUN   TestAccAWSLambdaFunction_expectFilenameAndS3Attributes
--- PASS: TestAccAWSLambdaFunction_expectFilenameAndS3Attributes (16.65s)
=== RUN   TestAccAWSLambdaFunction_envVariables
--- PASS: TestAccAWSLambdaFunction_envVariables (83.29s)
=== RUN   TestAccAWSLambdaFunction_encryptedEnvVariables
--- FAIL: TestAccAWSLambdaFunction_encryptedEnvVariables (66.91s)
	testing.go:513: Step 1 error: Check failed: Check 4/5 error: aws_lambda_function.lambda_function_test: Attribute 'environment.0.variables.foo' not found
=== RUN   TestAccAWSLambdaFunction_versioned
--- PASS: TestAccAWSLambdaFunction_versioned (37.98s)
=== RUN   TestAccAWSLambdaFunction_DeadLetterConfig
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfig (37.96s)
=== RUN   TestAccAWSLambdaFunction_DeadLetterConfigUpdated
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfigUpdated (52.98s)
=== RUN   TestAccAWSLambdaFunction_nilDeadLetterConfig
--- PASS: TestAccAWSLambdaFunction_nilDeadLetterConfig (15.72s)
=== RUN   TestAccAWSLambdaFunction_tracingConfig
--- PASS: TestAccAWSLambdaFunction_tracingConfig (52.04s)
=== RUN   TestAccAWSLambdaFunction_VPC
--- PASS: TestAccAWSLambdaFunction_VPC (35.89s)
=== RUN   TestAccAWSLambdaFunction_VPCUpdate
--- PASS: TestAccAWSLambdaFunction_VPCUpdate (51.68s)
=== RUN   TestAccAWSLambdaFunction_VPC_withInvocation
--- PASS: TestAccAWSLambdaFunction_VPC_withInvocation (57.75s)
=== RUN   TestAccAWSLambdaFunction_s3
--- PASS: TestAccAWSLambdaFunction_s3 (34.87s)
=== RUN   TestAccAWSLambdaFunction_localUpdate
--- PASS: TestAccAWSLambdaFunction_localUpdate (47.31s)
=== RUN   TestAccAWSLambdaFunction_localUpdate_nameOnly
--- PASS: TestAccAWSLambdaFunction_localUpdate_nameOnly (47.18s)
=== RUN   TestAccAWSLambdaFunction_s3Update_basic
--- PASS: TestAccAWSLambdaFunction_s3Update_basic (71.41s)
=== RUN   TestAccAWSLambdaFunction_s3Update_unversioned
--- PASS: TestAccAWSLambdaFunction_s3Update_unversioned (55.60s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_noRuntime
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_noRuntime (1.40s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_nodeJs43
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_nodeJs43 (36.50s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_python27
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_python27 (35.95s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_java8
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_java8 (37.37s)
=== RUN   TestAccAWSLambdaFunction_tags
--- PASS: TestAccAWSLambdaFunction_tags (67.01s)
=== RUN   TestAccAWSLambdaFunction_runtimeValidation_python36
--- PASS: TestAccAWSLambdaFunction_runtimeValidation_python36 (37.45s)
FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1309.176s
make: *** [testacc] Error 1

⎇  make testacc TEST=./aws TESTARGS='-run=TestAccAWSLambdaFunction_encryptedEnvVariables'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSLambdaFunction_encryptedEnvVariables -timeout 120m
=== RUN   TestAccAWSLambdaFunction_encryptedEnvVariables
--- PASS: TestAccAWSLambdaFunction_encryptedEnvVariables (78.43s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	78.474s
```

**Comment by loivis** [comment]
```
  make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAWSLambdaFunction_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccDataSourceAWSLambdaFunction_ -timeout 120m
=== RUN   TestAccDataSourceAWSLambdaFunction_basic
--- PASS: TestAccDataSourceAWSLambdaFunction_basic (41.44s)
=== RUN   TestAccDataSourceAWSLambdaFunction_version
--- PASS: TestAccDataSourceAWSLambdaFunction_version (41.08s)
=== RUN   TestAccDataSourceAWSLambdaFunction_alias
--- PASS: TestAccDataSourceAWSLambdaFunction_alias (42.43s)
=== RUN   TestAccDataSourceAWSLambdaFunction_vpc
--- PASS: TestAccDataSourceAWSLambdaFunction_vpc (37.63s)
=== RUN   TestAccDataSourceAWSLambdaFunction_environment
--- PASS: TestAccDataSourceAWSLambdaFunction_environment (40.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	203.260s
```

**Comment by Puneeth-n** [comment]
@lovis see my issue #2959

**Comment by loivis** [comment]
@Puneeth-n thanks. good to know there is a fix.

**Comment by Puneeth-n** [comment]
@loivis Nope. No fix.

**Comment by t-pascal** [comment]
Plus one for this feature. I have recently needed to import lambda function qualified alias as a data source.

**Comment by kazeshini178** [comment]
Any idea when this would be merged into the master branch/released would be really useful?

File: aws/resource_aws_lambda_function.go [hunk 1]
```diff
@@ -161,6 +161,11 @@ func resourceAwsLambdaFunction() *schema.Resource {
 				Optional: true,
 				Computed: true,
 			},
+			"source_code_size": {
```
Comment (bflad) [inline aws/resource_aws_lambda_function.go]: Two things:
* This does not look configurable -- we should remove `Optional: true`
* Its not documented in the resource documentation

File: aws/resource_aws_lambda_function.go [hunk 2]
```diff
@@ -385,12 +387,19 @@ func resourceAwsLambdaFunctionCreate(d *schema.ResourceData, meta interface{}) e
...
 	conn := meta.(*AWSClient).lambdaconn
 
-	log.Printf("[DEBUG] Fetching Lambda Function: %s", d.Id())
-
 	params := &lambda.GetFunctionInput{
 		FunctionName: aws.String(d.Get("function_name").(string)),
 	}
 
+	// qualifier for lambda function data source
+	qualifier, qualifierExistance := d.GetOk("qualifier")
```
Comment (bflad) [inline aws/resource_aws_lambda_function.go]: Generally speaking, if there starts being logic like this in a shared resource/data source read function, we will want to not reuse the function.

File: aws/resource_aws_lambda_function.go [hunk 3]
```diff
@@ -416,19 +437,23 @@ func resourceAwsLambdaFunctionRead(d *schema.ResourceData, meta interface{}) err
...
+	d.Set("source_code_hash", function.CodeSha256)
+	d.Set("source_code_size", function.CodeSize)
 
-	config := flattenLambdaVpcConfigResponse(function.VpcConfig)
-	log.Printf("[INFO] Setting Lambda %s VPC config %#v from API", d.Id(), config)
-	vpcSetErr := d.Set("vpc_config", config)
-	if vpcSetErr != nil {
-		return fmt.Errorf("Failed setting vpc_config: %s", vpcSetErr)
+	vpcConfig := flattenLambdaVpcConfigResponse(function.VpcConfig)
+	if vpcConfig != nil {
```
Comment (bflad) [inline aws/resource_aws_lambda_function.go]: This will likely break module/output usage. Why was this changed?

File: aws/resource_aws_lambda_function.go [hunk 4]
```diff
@@ -416,19 +437,23 @@ func resourceAwsLambdaFunctionRead(d *schema.ResourceData, meta interface{}) err
...
+			return fmt.Errorf("Failed setting vpc_config: %s", err)
+		}
 	}
 
-	d.Set("source_code_hash", function.CodeSha256)
-
-	if err := d.Set("environment", flattenLambdaEnvironment(function.Environment)); err != nil {
-		log.Printf("[ERR] Error setting environment for Lambda Function (%s): %s", d.Id(), err)
+	environment := flattenLambdaEnvironment(function.Environment)
+	if environment != nil {
```
Comment (bflad) [inline aws/resource_aws_lambda_function.go]: Same here. This will likely break module/output usage. Why was this changed?

File: aws/resource_aws_lambda_function.go (hunk 3 above)
Comment (loivis) [reply aws/resource_aws_lambda_function.go]: I can't really recall what I was thinking at that time 🤔

File: aws/resource_aws_lambda_function.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_lambda_function.go]: We can always adjust this later if necessary as its probably fine the way it currently is. 👍

**Comment by bflad** [comment]
This has been released in [version 1.16.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1160-april-25-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
